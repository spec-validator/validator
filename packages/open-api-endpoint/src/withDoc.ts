import { declareField, Field } from '@spec-validator/validator/core'
import { Any } from '@spec-validator/utils/util-types'

type Example<T extends Any> = {
  value: T,
  summary: string
}

type Doc<T extends Any> = {
  description?: string,
  format?: string,
  examples?: Record<string, Example<T>>
}

export interface WithDoc<T extends Any, F extends Field<T>> extends Field<T> {
  readonly innerField: F,
  readonly doc: Doc<T>
}

export default declareField('@spec-validator/fields.WithDoc', <T extends Any, F extends Field<T>> (
  innerField: F,
  doc: Doc<T>
): WithDoc<T, F> => ({
    innerField,
    doc,
    validate: (it) => innerField.validate(it),
    serialize: (it) => innerField.serialize(it),
  }))
