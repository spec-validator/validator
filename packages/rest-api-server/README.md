# @spec-validator/rest-api-server

The library provides a set of types that specifies HTTP protocol.

Aside of the types it also implements the core logic of an HTTP
server that on the one hand routinely handles incoming requests
and does the routing in a similar fashion `express` does it
and on the other hand leverages type specifications to validate
the incoming requests.

## Routes

The core building block of a server is a route.

The route is a combination of essentially 3 things:

- request object specification
- response object specification
- handler that gets a valid request object as its only parameter and
  returns a valid response object

```ts
import { route, Route } from '@spec-validator/rest-api-server'
import { expectType } from '@spec-validator/test-utils/expectType'
import {
  segmentField as $, choiceField, numberField, stringField,
  unionField, constantField,
} from '@spec-validator/validator/fields'

const routeObject = route({
  request: {
    method: constantField('GET'),
    headers: {
      headerKey: stringField(),
    },
    pathParams: $._('pathKey', stringField()),
  },
  response: unionField(
    {
      statusCode: constantField(201),
      body: numberField(),
    }, {
      statusCode: constantField(202),
      body: choiceField('one', 'two'),
      headers: {
        headerKey: stringField(),
    },
  }),
}).handler(async (request) => ({
  statusCode: 202,
  body: 'one',
  headers: {
    headerKey: request.pathParams.pathKey,
  },
}))

expectType<Route, typeof routeObject>(true)
```

Note, the routes do not handle trafic in any meaningful way,
they are just a specification for the handlers.

### Fluent API for the most common case

There is a fluent API aiming to simplify route configuration in the
most common case (one 200ish response per handler, known status
codes for various request methods).

```ts
import { createRouteCollection  } from '@spec-validator/rest-api-server'

const itemSpec = {
  title: stringField(),
  description: stringField(),
}

// 'routes' is a list that effectively will aggregate the routes
// defined via a '_' helper object
const _ = createRouteCollection()

// As an alternative you may import '_' singleton object that is
// not linked to any routes collection. In that case you would need
// to add routes to the collections explicitly via 'routes.push'

/**
import { _ } from '@spec-validator/rest-api-server'

const routes: Route[] = []

routes.push(_.GET(...).spec(...).handler(...))
*/

const getRoute = _.GET($._('/items')).spec({
  response: {
    body: [itemSpec],
  },
}).handler(async () => ({
  body: [
    {
      title: 'Item N',
      description: 'Description',
    },
  ],
}))

expectType<Route, typeof getRoute>(true)

const postRoute = _.POST($._('/items')).spec({
  request: {
    body: itemSpec,
  },
  response: {
    body: numberField(),
    headers: {
      title: stringField(),
    },
  },
}).handler(async () => ({
  body: 42,
  headers: {
    title: 'Foo',
  },
}))

expectType<Route, typeof postRoute>(true)
```

## Server object

For any arbitrary set of routes it is possible to join those with
extra parameters to craft server configs:

```ts
import { createServer } from '@spec-validator/rest-api-server'

import http from 'http'

const server = createServer({
  baseUrl: 'http://localhost:8080',
  routes: _.routes
})

expectType<http.Server & { serve: () => http.Server }, typeof server>(true)
```

## Running the app

To actually run the server and handle trafic using it,
run `serve` method:

```ts #ignore

server.serve()

```

Once the app is started you may run the following curl command to obtain the
payload from the endpoint:

```
curl -X GET "http://localhost:8080/items"
```

Or post a new entry:

```
curl -X POST "http://localhost:8080/items" -H  "Content-Type: application/json" \
  -d "{\"title\":\"string\",\"description\":\"string\"}"
```

The following call will produce a 400 (validation) error:

```
curl -X POST "http://localhost:8080/items" -H  "Content-Type: application/json" \
  -d "{\"title\":true,\"description\":\"string\"}"
```

## Cheat sheet

Response with cookies:

```ts
import { headerObjectField } from '@spec-validator/rest-api-server/fields'

const withCookies = _.GET($._('/with-cookies')).spec({
  response: {
    body: [itemSpec],
    headers: {
      cookies: headerObjectField({
        cookieStr: stringField(),
        cookieNum: numberField()
      })
    }
  },
}).handler(async () => ({
  body: [
    {
      title: 'Item N',
      description: 'Description',
    },
  ],
  headers: {
    cookies: {
      cookieStr: 'foo',
      cookieNum: 42
    }
  }
}))

expectType<Route, typeof withCookies>(true)
```

Request with authorization:

```ts
const withAuthorization = _.POST($._('/with-cookies')).spec({
  request: {
    headers: {
      authorization: $._('type', stringField())._(' ')._('credentials', stringField())
    }
  }
}).handler(async () => undefined)


expectType<Route, typeof withAuthorization>(true)
```
