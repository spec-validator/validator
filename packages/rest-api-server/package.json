{
  "name": "@spec-validator/rest-api-server",
  "version": "0.0.1",
  "dependencies": {
    "@spec-validator/utils": "0.0.1",
    "@spec-validator/validator": "0.0.1",
    "qs": "^6.9.4"
  },
  "devDependencies": {
    "@spec-validator/test-utils": "0.0.1",
    "@types/express": "^4.17.9",
    "@types/qs": "^6.9.5",
    "@types/supertest": "^2.0.10",
    "express": "^4.17.1",
    "supertest": "^6.0.1"
  }
}
