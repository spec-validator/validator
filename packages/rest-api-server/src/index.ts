export { ServerConfig } from './handler'
export { Route, route } from './route'
export { serve, withMethod, _ } from './server'
