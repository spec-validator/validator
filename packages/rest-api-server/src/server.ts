import http from 'http'
import { URL } from 'url'
import { route, StringMapping } from './route'

import { ServerConfig, handle } from './handler'
import { HtmlSerialization, JsonSerialization } from './serialization'
import { Route, RequestSpec, ResponseSpec } from './route'

import { TypeHint } from '@spec-validator/validator'
import { constantField, $ } from '@spec-validator/validator/fields'
import { WithoutOptional } from 'utils/src/util-types'
import { ConstantField } from '@spec-validator/validator/fields/constantField'
import { Field } from '@spec-validator/validator/core'

type RequestSpecMethod = Omit<RequestSpec, 'method' | 'pathParams'>

type ResponseSpecMethod = Omit<ResponseSpec, 'statusCode'>

type PathSpec<PathParams extends StringMapping> = typeof $ & Field<PathParams>

// Make it fluid API - to make things work with autocomplete
export const withMethod = <
  Method extends string,
  OkStatusCode extends number
> (method: Method, okStatusCode: OkStatusCode) =>
  <PathParams extends StringMapping = StringMapping> (pathParams: PathSpec<PathParams>): {
    spec:  <
      ReqSpec extends RequestSpecMethod = RequestSpecMethod,
      RespSpec extends ResponseSpecMethod = ResponseSpecMethod,
      > (spec: {
      request?: ReqSpec,
      response?: RespSpec
    }) => ({
      handler: (
        handler: (request: WithoutOptional<TypeHint<
          ReqSpec
          & {
              readonly method: ConstantField<Method>,
              readonly pathParams: PathSpec<PathParams>
            }
        >>
      ) => Promise<
        WithoutOptional<TypeHint<RespSpec>>
      >) => Route
    })
  } => ({
      spec: (spec) => ({ handler: ( handler ) =>
        route.spec({
          request: {
            ...(spec.request || {}),
            method: constantField(method),
            pathParams,
          },
          response: {
            ...(spec.response || {}),
            statusCode: constantField(okStatusCode),
          },
        }).handler(
          async (req) => ({ ...((await handler(req as any)) as any), statusCode: okStatusCode })
        ),
      }),
    })

export const _ = {
  GET: withMethod('GET', 200),
  HEAD: withMethod('HEAD', 200),
  POST: withMethod('POST', 201),
  PUT: withMethod('PUT', 204),
  DELETE: withMethod('DELETE', 204),
  PATCH: withMethod('PATCH', 204),
}

export const DEFAULT_SERVER_CONFIG: ServerConfig = {
  baseUrl: 'http://localhost:8000',
  serializationFormats: [new JsonSerialization(), new HtmlSerialization()],
  encoding: 'utf-8',
  frameworkErrorStatusCode: 503,
  appErrorStatusCode: 500,
  reportError: (error: unknown) => {
    console.error(error)
    return Promise.resolve(undefined)
  },
  routes: [],
}

const mergeServerConfigs = (
  serverConfig: Partial<ServerConfig>
): ServerConfig => ({
  ...DEFAULT_SERVER_CONFIG,
  ...serverConfig,
})

const SUPPORTED_PROTOCOLS= {
  'http': 80,
  'https': 443,
} as const

const supportedProtocols = new Set(Object.keys(SUPPORTED_PROTOCOLS))

const verifyProtocol = (protocol: string): protocol is keyof typeof SUPPORTED_PROTOCOLS =>
  supportedProtocols.has(protocol)

const getPort = (baseUrl: string): number => {
  const url = new URL(baseUrl)
  // Node's protocol has a trailing column
  const protocol = url.protocol.replace(/:$/, '')
  if (!verifyProtocol(protocol)) {
    throw `Protocol ${url.protocol} is not supported`
  }
  return url.port ? Number.parseInt(url.port) : SUPPORTED_PROTOCOLS[protocol]
}

export const createServer = (
  config: Partial<ServerConfig>,
): http.Server => http.createServer(handle.bind(null, mergeServerConfigs(config)))

export const serve = (
  config: Partial<ServerConfig>,
): void => {
  const merged = mergeServerConfigs(config)
  http.createServer(handle.bind(null, merged)).listen(getPort(merged.baseUrl))
}
