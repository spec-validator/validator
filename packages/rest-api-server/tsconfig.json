{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "noEmit": true,
    "paths": {
      "@spec-validator/validator/*": ["../validator/src/*"],
      "@spec-validator/validator": ["../validator/src/index.ts"]
    }
  },
  "include": ["src/**/*.ts"]
}
