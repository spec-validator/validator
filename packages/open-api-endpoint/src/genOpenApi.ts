import { OpenAPIV3 as OpenAPI } from 'openapi-types'

import { ServerConfig, Route } from '@spec-validator/rest-api-server'
import { $, optional } from '@spec-validator/validator/fields'
import { GetRepresentation } from '@spec-validator/validator/registry'

import getFieldSchema from './schemaRegistry'
import { Any, ConstructorArgs } from '@spec-validator/utils/util-types'
import { isResponsesSpec, ResponseSpec } from '@spec-validator/rest-api-server/route'
import { withoutOptional } from '@spec-validator/utils/utils'
import { isFieldSpec, SpecUnion } from '@spec-validator/validator/core'
import { Json } from '@spec-validator/utils/Json'

const mergeValues = (pairs: [a: string, b: OpenAPI.PathItemObject][]): Record<string, OpenAPI.PathItemObject> => {
  const record: Record<string, OpenAPI.PathItemObject>  = {}
  pairs.forEach(([a, b]) => {
    record[a] = {
      ...(record[a] || {}),
      ...b,
    }
  })
  return record
}

export type Info = {
  title: string,
  version: string
}

export type WithInfo = {
  info?: Info
}

export const DEFAULT_INFO: Info = {
  title: 'Test',
  version: '0.0.1',
}

const getDescription = (field: unknown): string | undefined =>
  (field as { description?: string }).description?.toString()

const isRequired = (spec: SpecUnion<unknown>) => {
  if (isFieldSpec(spec)) {
    return (spec as any).type !== optional.type
  } else {
    return !!spec
  }
}

type ParameterType = 'query' | 'path'

class OpenApiGenerator {

  getSchema: GetRepresentation

  constructor(
    readonly config: ServerConfig & WithInfo,
    getSchema: GetRepresentation = getFieldSchema
  ) {
    this.getSchema = (field) => withoutOptional(getSchema(field))
  }

  createOpenApiSpec = (): OpenAPI.Document => withoutOptional({
    openapi: '3.0.3',
    info: {
      ...DEFAULT_INFO,
      ...this.config.info,
    },
    servers: [
      {
        url: this.config.baseUrl,
      },
    ],
    paths: mergeValues(this.config.routes.map(this.createPath)),
  })

  createParameterBaseObject = (
    spec: SpecUnion<unknown>
  ): OpenAPI.ParameterBaseObject => withoutOptional({
    description: getDescription(spec),
    required: isRequired(spec),
    schema: this.getSchema(spec),
  })

  createParameter = (
    type: ParameterType,
    name: string,
    spec: SpecUnion<unknown>
  ): OpenAPI.ParameterObject => withoutOptional({
    name: name,
    in: type,
    ...this.createParameterBaseObject(spec),
  })

  specToParams = (
    type: ParameterType,
    spec?: Record<string, SpecUnion<unknown>>
  ): OpenAPI.ParameterObject[] => Object.entries(spec || {}).map(
    ([name, field]) => this.createParameter(type, name, field)
  )

  createContentObject = (data: SpecUnion<unknown>): {
    [media: string]: OpenAPI.MediaTypeObject
  } => Object.fromEntries(this.config.serializationFormats.map(
    it => [it.mediaType, {
      schema: this.getSchema(data),
    }]
  ))

  createResponseObject = (response: ResponseSpec): OpenAPI.ResponseObject => withoutOptional({
    headers: response.headers && Object.fromEntries(Object.entries(response.headers).map(([name, value]) => [
      name,
      this.createParameterBaseObject(value),
    ])),
    description: getDescription(response) || '',
    content: response.body && this.createContentObject(response.body),
  })

  createRequestBodyObject = (data: SpecUnion<Any>): OpenAPI.RequestBodyObject => withoutOptional({
    content: this.createContentObject(data),
    required: isRequired(data),
  })

  createResponses = (spec: Route['response']) => {
    const result: Record<string, OpenAPI.ResponseObject> = {}
    if (isResponsesSpec(spec)) {
      spec.variants.forEach((it) => {
        result[it.statusCode.constant.toString()] = this.createResponseObject(it)
      })
    } else {
      result[spec.statusCode.constant.toString()] = this.createResponseObject(spec)
    }
    return result
  }

  createOperationObject = (route: Route): OpenAPI.OperationObject => withoutOptional({
    parameters: [
      ...this.specToParams('query', route.request.queryParams),
      ...this.specToParams('path', route.request.pathParams.getObjectSpec()),
    ],
    requestBody: route.request.body && this.createRequestBodyObject(route.request.body),
    responses: this.createResponses(route.response),
  })

  createPathString = (pathParams: typeof $): string =>
    pathParams.getSegments().map(it => it.field ? `{${it.key}}` : it.key || '').join('')

  createPath = (route: Route): [string, OpenAPI.PathItemObject] => [this.createPathString(route.request.pathParams), {
    [(route.request.method.constant as string).toLowerCase()]: this.createOperationObject(route),
  }]

}

export default (...params: ConstructorArgs<typeof OpenApiGenerator>): Json =>
  new OpenApiGenerator(...params).createOpenApiSpec() as unknown as Json
