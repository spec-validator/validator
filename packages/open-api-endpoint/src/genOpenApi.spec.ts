import { _, Route, route } from '@spec-validator/rest-api-server'
import {
  segmentField as $, stringField, booleanField, numberField, arrayField, unionField, constantField,
} from '@spec-validator/validator/fields'
import genOpenApi from './genOpenApi'
import { DEFAULT_SERVER_CONFIG } from '@spec-validator/rest-api-server/server'
import withDoc from './withDoc'

test('fullRoute', () => {
  const routes: Route[] = [
    route(
      {
        request: {
          pathParams: $._('/multi-response'),
          method: constantField('GET'),
        },
        response: unionField({
          statusCode: constantField(201),
        }, {
          statusCode: constantField(202),
        }),
      }
    ).handler(
      async () => Promise.resolve({
        statusCode: 201,
      })
    ),
    _.POST($._('/items')).spec(
      {
        request: {
          body: {
            title: stringField(/.*/),
            count: numberField(),
          },
        },
      }
    ).handler(
      async () => Promise.resolve(undefined)
    ),
    _.GET($._('/item/')._('id', stringField())).spec(
      {
        request: {
          headers: {
            key: withDoc(numberField(), {
              description: 'key header',
              examples: {
                sampleKey: {
                  value: 13 as number,
                  summary: 'Sample value',
                },
              },
            }),
          },
          queryParams: {
            flag: booleanField(),
          },
        },
        response: {
          headers: {
            key: withDoc(numberField(), {
              description: 'key header',
              examples: {
                sampleKey: {
                  value: 13 as number,
                  summary: 'Sample value',
                },
              },
            }),
          },
          body: {
            items: arrayField(numberField()),
          },
        },
      }
    ).handler(
      async () => ({
        statusCode: 200,
        headers: {
          key: 13,
        },
        body: {
          items: [1, 2, 3],
        },
      })
    ),
  ]

  expect(genOpenApi(
    {
      ...DEFAULT_SERVER_CONFIG,
      routes,
      info: {
        title: 'Test',
        version: '1.0.0',
      },
    })).toMatchSnapshot()

})
