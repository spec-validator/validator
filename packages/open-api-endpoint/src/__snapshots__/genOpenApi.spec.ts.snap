// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fullRoute 1`] = `
Object {
  "info": Object {
    "title": "Test",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/item/{id}": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "flag",
            "required": true,
            "schema": Object {
              "type": "boolean",
            },
          },
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "pattern": ".*",
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "items": Object {
                      "items": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "items",
                  ],
                  "type": "object",
                },
              },
              "text/html": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "items": Object {
                      "items": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "items",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
            "headers": Object {
              "key": Object {
                "required": true,
                "schema": Object {
                  "description": Object {
                    "description": "key header",
                    "examples": Object {
                      "sampleKey": Object {
                        "summary": "Sample value",
                        "value": 13,
                      },
                    },
                  },
                  "minimum": 0,
                  "type": "integer",
                },
              },
            },
          },
        },
      },
    },
    "/items": Object {
      "post": Object {
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "count": Object {
                    "minimum": 0,
                    "type": "integer",
                  },
                  "title": Object {
                    "pattern": ".*",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                  "count",
                ],
                "type": "object",
              },
            },
            "text/html": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "count": Object {
                    "minimum": 0,
                    "type": "integer",
                  },
                  "title": Object {
                    "pattern": ".*",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                  "count",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "201": Object {
            "description": "",
          },
        },
      },
    },
    "/multi-response": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "201": Object {
            "description": "",
          },
          "202": Object {
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [
    Object {
      "url": "http://localhost:8000",
    },
  ],
}
`;
